src/mlpack/tests/sdp_primal_dual_test.cpp

```diff
@@ -557,92 +558,141 @@ BOOST_AUTO_TEST_CASE(CorrelationCoeffToySdp)
+ *   Kilian Weinberger and Lawrence Saul. CVPR 04.
+ *   http://repository.upenn.edu/cgi/viewcontent.cgi?article=1000&context=cis_papers
+ *
+ * Here is another paper which clarifies more specific SDP formulation of MVU problem.
+ *
+ *   Scalable semidefinite manifold learning
+ *   Nikolaos Vasiloglou, Alexander G. Gray, David V. Anderson.
+ *   2008 IEEE Workshop on Machine Learning for Signal Processing
+ *
+ * @param origData origDim x numPoints
+ * @param numNeighbors
+ */
+static inline SDP<arma::sp_mat> ConstructMvuSDP(const arma::mat& origData,
+                                              size_t numNeighbors)
+{
+  const size_t numPoints = origData.n_cols;
+
+  assert(numNeighbors <= numPoints);
+
+  arma::Mat<size_t> neighbors;
+  arma::mat distances;
+  KNN knn(origData);
+  knn.Search(numNeighbors, neighbors, distances);
+
+  // how to construct the matrices A, b, and C? Plz refer to the above paper.
+  std::set<std::pair<int, int>> neighbors_set;
+  for (size_t i = 0; i < neighbors.n_cols; ++i)
+  {
+    for (size_t j = 0; j < numNeighbors; ++j)
+    {
+      size_t x = i, y = neighbors(j, i);
+      if (x > y)
+      {
+        x = neighbors(j, i);
+        y = i;
+      }
+
+      if (neighbors_set.find(std::make_pair(x, y)) != neighbors_set.end())
+      {
+        continue;
+      }
+
+      neighbors_set.insert(std::make_pair(x, y));
+    }
+  }
+
+  // SDP<arma::sp_mat> sdp(numPoints, numNeighbors * numPoints, 1);
+  SDP<arma::sp_mat> sdp(numPoints, neighbors_set.size(), 1);
+  sdp.C().eye(numPoints, numPoints);
+  sdp.C() *= -1;
+  sdp.DenseA()[0].ones(numPoints, numPoints);
+  sdp.DenseB()[0] = 0;
+  neighbors_set.clear();
+  size_t _index = 0;
+
+  for (size_t i = 0; i < neighbors.n_cols; ++i)
+  {
+    for (size_t j = 0; j < numNeighbors; ++j)
+    {
+      // The set is used to filter duplicate neighbor pair.
+      // Then the linear indenpence of the constraints can be satisfied.
+      size_t x = i, y = neighbors(j, i);
+      if (x > y)
+      {
+        x = neighbors(j, i);
+        y = i;
+      }
+      if (neighbors_set.find(std::make_pair(x, y)) != neighbors_set.end())
+      {
+        continue;
+      }
+      neighbors_set.insert(std::make_pair(x, y));
+
+      // This is the index of the constraint.
+      const size_t index = _index;
+      _index++;
+      arma::sp_mat& aRef = sdp.SparseA()[index];
+      aRef.zeros(numPoints, numPoints);
+
+      // A_ij(i, i) = 1.
+      aRef(i, i) = 1;
+
+      // A_ij(i, j) = -1.
+      aRef(i, neighbors(j, i)) = -1;
+
+      // A_ij(j, i) = -1.
+      aRef(neighbors(j, i), i) = -1;
+
+      // A_ij(j, j) = 1.
+      aRef(neighbors(j, i), neighbors(j, i)) = 1;
+
+      // The constraint b_ij is the distance between these two points.
+      sdp.SparseB()[index] = distances(j, i);
+    }
+  }
+
+  return sdp;
+}
+
+/**
+ * Maximum variance unfolding
+ *
+ * Test doesn't work, because the constraint matrices are not linearly
+ * independent.
```

Comment:
It seems like the test does work now, so I think we can remove this comment.

Reply:
It may be a starting point issue.  You might try running the test many times with different random seeds and seeing how different the results are, and how often it fails/succeeds.  (Note that if you use `math::RandomSeed()`, you'll have to comment out the #ifdef in src/mlpack/core/math/random.hpp.)

Reply:
What do you mean "comment out the #ifdef in src/mlpack/core/math/random.hpp"? Can I just use `mlpack::math::RandomSeed((size_t) std::time(NULL))` ?

Reply:
Ah, sorry, this comment was incorrect on my end.  If we were writing a test for any of the command-line bindings, using different random seeds is disabled via the `#ifdef`s in `random.hpp` (since some bindings will always try to set a random seed, and we don't want that to happen for reproducibility).  But since you are not working with any test in `main_tests/`, the issue doesn't apply here---`sdp_primal_dual_test.cpp` is not built with `BINDING_TYPE` set to anything at all.

Sorry for the confusion!